### DISCUSSION TITLE
3.3: inappropriately hard/missing info?

### DISCUSSION THREAD

**Discussion by jalexchapman** [TEALSK12/introduction-to-computer-science#430] [General]
Date math and civil time are famously pernicious things, where novice developers think they can just gin up their own implementation because how hard can it be, and they promptly learn why we have libraries for this stuff. And true to form, this lab is a lot harder than it looks. I had the pleasure of working with a longtime test owner of the calendar in Windows and IIRC she gave a pretty interesting talk on testing that used it as an example. Look up "inter gravissimas" (the papal bull that converted from Julian to Gregorian calendars) and think about the test cases that adds. The Windows calendar APIs do handle 1582 (and years preceding) correctly.

- The "is leap year" _requires_ the `x mod y` block, which has not been introduced yet, is not mentioned in lab notes or lecture, and is not the thing most students will be able to intuitively apply based on their math training. The instructions should explain and apply the `x mod y` block explicitly.
- Unless I'm missing something pretty major, the `day x of year y` block is impossible without duplicating code in previous blocks. Since 3.4 is the lab that introduces predicate and reporter blocks, the "is leap year" and "days in month" blocks are both Command blocks that just `say` their results, so you can't make use of them inside of your `day x of year y` block, but you absolutely need their logic. This heavy reuse is likely to be discouraging to students, and the counting off of remaining days (but look out for leap years!) is a pretty tricky algorithm for where students are at this point. I would recommend this as a bonus problem _at best_, again, unless I'm missing a pretty major trick.
- The "what day of the week were you born" bonus problem hint points at a very long Wikipedia article with a _mess_ of algorithms. You might point more specifically at Rata Die and note that January 1 1967 was a Sunday. But students would still have to include a flipped version of `days in month` and a duplicate of `is leap year` to do this without the benefit of the reporter/predicate blocks the next lab introduces. This is unreasonably hard/tedious for the level of this course, especially when the next lab is about to introduce a tool that makes the problem vastly less annoying. I'd drop this entirely and treat `day x of year y` as the bonus content, especially since the core algorithm isn't much different than the one you use there.

**Comment by jdonwells** [comment]
I also think it needs to be replaced. One thing you didn't mention is how much easier many of these are with lists. 

I actually replaced the 4.1 lab with a redux of some of 3.3 but with lists. I give them a leap year predicate as part of the starter code, among other things.

**Reply by jpfasano** [reply]
@jdonwells Can you share your replacement of Lab 4.1?  I'm starting to look at 4.1 where students are asked to "write an algorithm for adding new Solar System discoveries to a collection of older ones."  I'm trying to figure out how to scaffold this assignment and suspect revisiting 3.3 with lists is probably more accessible for my students.  Thank you.

**Comment by jdonwells** [comment]
Slide show. https://docs.google.com/presentation/d/13zRxzuondpTRv2JzXM0t1rv_fL6iwxu7_cVTXPNZAGA/edit#slide=id.g159b1ee3b43_0_39

Lab.
https://docs.google.com/document/d/1Dm66YDhgn7SBZyvIFx7T-1he4QWkqL4UJZ00glQ6AkQ/edit

**Reply by jpfasano** [reply]
@jdonwells Awesome.  I expect to use your version of 4.1.  Thank you.

**Comment by Mykel19XX** [comment]
THANK YOU! This is the third year I've taught this course and I've always disliked the way this unit is constructed. If we show students how to make custom blocks in 3.1, why is there a massive detour to physics simulation in 3.2 that doesn't make use of custom blocks? Then 3.3 has the issues you pointed out. Just changing the order of the lessons would be a massive improvement. Lab 3.1 first, Part 1 of Lab 3.3 second, Lab 3.4 third, Part 2 of Lab 3.3 but only exercises 1 and 2 forth, and finally Lab 3.2, but altered to make use of custom blocks and serve as a launching point for the platformer lab.
